File: go/test/air_temperature_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/realtime-weather-sdk"
	"github.com/voxgig-sdk/realtime-weather-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestAirTemperatureEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.AirTemperature(nil)
		if ent == nil {
			t.Fatal("expected non-nil AirTemperatureEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := air_temperatureBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "air_temperature." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set REALTIMEWEATHER_TEST_AIR_TEMPERATURE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		airTemperatureRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.air_temperature", setup.data)))
		var airTemperatureRef01Data map[string]any
		if len(airTemperatureRef01DataRaw) > 0 {
			airTemperatureRef01Data = core.ToMapAny(airTemperatureRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = airTemperatureRef01Data

		// LIST
		airTemperatureRef01Ent := client.AirTemperature(nil)
		airTemperatureRef01Match := map[string]any{
			"collection_id": setup.idmap["collection01"],
		}

		airTemperatureRef01ListResult, err := airTemperatureRef01Ent.List(airTemperatureRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, airTemperatureRef01ListOk := airTemperatureRef01ListResult.([]any)
		if !airTemperatureRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", airTemperatureRef01ListResult)
		}

	})
}

func air_temperatureBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "air_temperature", "AirTemperatureTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read air_temperature test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse air_temperature test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"air_temperature01", "air_temperature02", "air_temperature03", "collection01", "collection02", "collection03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("REALTIMEWEATHER_TEST_AIR_TEMPERATURE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"REALTIMEWEATHER_TEST_AIR_TEMPERATURE_ENTID": idmap,
		"REALTIMEWEATHER_TEST_LIVE":      "FALSE",
		"REALTIMEWEATHER_TEST_EXPLAIN":   "FALSE",
		"REALTIMEWEATHER_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["REALTIMEWEATHER_TEST_AIR_TEMPERATURE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["REALTIMEWEATHER_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["REALTIMEWEATHER_APIKEY"],
			},
			extra,
		})
		client = sdk.NewRealtimeWeatherSDK(core.ToMapAny(mergedOpts))
	}

	live := env["REALTIMEWEATHER_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["REALTIMEWEATHER_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/AirTemperatureEntityTest.php
<?php
declare(strict_types=1);

// AirTemperature entity test

require_once __DIR__ . '/../realtimeweather_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class AirTemperatureEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = RealtimeWeatherSDK::test(null, null);
        $ent = $testsdk->AirTemperature(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = air_temperature_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "air_temperature." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set REALTIMEWEATHER_TEST_AIR_TEMPERATURE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $air_temperature_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.air_temperature")));
        $air_temperature_ref01_data = null;
        if (count($air_temperature_ref01_data_raw) > 0) {
            $air_temperature_ref01_data = Helpers::to_map($air_temperature_ref01_data_raw[0][1]);
        }

        // LIST
        $air_temperature_ref01_ent = $client->AirTemperature(null);
        $air_temperature_ref01_match = [
            "collection_id" => $setup["idmap"]["collection01"],
        ];

        [$air_temperature_ref01_list_result, $err] = $air_temperature_ref01_ent->list($air_temperature_ref01_match, null);
        $this->assertNull($err);
        $this->assertIsArray($air_temperature_ref01_list_result);

    }
}

function air_temperature_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/air_temperature/AirTemperatureTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = RealtimeWeatherSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["air_temperature01", "air_temperature02", "air_temperature03", "collection01", "collection02", "collection03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("REALTIMEWEATHER_TEST_AIR_TEMPERATURE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "REALTIMEWEATHER_TEST_AIR_TEMPERATURE_ENTID" => $idmap,
        "REALTIMEWEATHER_TEST_LIVE" => "FALSE",
        "REALTIMEWEATHER_TEST_EXPLAIN" => "FALSE",
        "REALTIMEWEATHER_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["REALTIMEWEATHER_TEST_AIR_TEMPERATURE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["REALTIMEWEATHER_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["REALTIMEWEATHER_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new RealtimeWeatherSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["REALTIMEWEATHER_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["REALTIMEWEATHER_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
